syntax = "proto3";

package tickets;

service TicketsService {
    rpc CreateTicket(CreateTicketRequest) returns (TicketResponse);
    rpc GetTicket(GetTicketRequest) returns (TicketResponse);
    rpc UpdateTicket(UpdateTicketRequest) returns (TicketResponse);
    rpc ListTickets(ListTicketsRequest) returns (TicketsListResponse);
    rpc AddComment(AddCommentRequest) returns (CommentResponse);
    rpc ListComments(ListCommentsRequest) returns (CommentsListResponse);
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

message Ticket {
    int32 id = 1;
    string title = 2;
    string description = 3;
    string category = 4;
    string status = 5;  // open, in_progress, resolved, closed
    int32 created_by = 6;
    string created_by_name = 7;
    int32 assigned_to = 8;
    string assigned_to_name = 9;
    int32 priority = 10;  // 1-5
    string created_at = 11;
    string updated_at = 12;
}

message Comment {
    int32 id = 1;
    int32 ticket_id = 2;
    int32 user_id = 3;
    string user_name = 4;
    string content = 5;
    string created_at = 6;
}

message CreateTicketRequest {
    string title = 1;
    string description = 2;
    string category = 3;
    int32 created_by = 4;
    int32 priority = 5;
}

message GetTicketRequest {
    int32 id = 1;
}

message UpdateTicketRequest {
    int32 id = 1;
    string status = 2;
    int32 assigned_to = 3;
    int32 priority = 4;
}

message ListTicketsRequest {
    int32 page = 1;
    int32 page_size = 2;
    int32 created_by = 3;
    int32 assigned_to = 4;
    string status = 5;
    string category = 6;
}

message AddCommentRequest {
    int32 ticket_id = 1;
    int32 user_id = 2;
    string content = 3;
}

message ListCommentsRequest {
    int32 ticket_id = 1;
}

message TicketResponse {
    Ticket ticket = 1;
    string message = 2;
}

message TicketsListResponse {
    repeated Ticket tickets = 1;
    int32 total_count = 2;
}

message CommentResponse {
    Comment comment = 1;
    string message = 2;
}

message CommentsListResponse {
    repeated Comment comments = 1;
    int32 total_count = 2;
}

message HealthCheckRequest {}
message HealthCheckResponse {
    string status = 1;
    string version = 2;
    string timestamp = 3;
}

